src/Entity/MetadataDisplayEntity.php

```diff
@@ -473,95 +507,57 @@ public function getTwigVariablesUsed() {
+      ) {
+        $variable_key = $nodes->getNode('node')->getAttribute('name')
+                        . '.' . $node->getAttribute('value');
       }
-      elseif ($node instanceof GetAttrExpression) {
-        // The array_unique/array_merge pattern below can't be used here because
-        // they don't work on multidimensional arrays. Instead the lines from
-        // $all_variables is passed to $variables below, and then the merged
-        // lines replace the old values and pass them thru the recursion.
-        // @todo: Look deeper into this and find a simpler, cleaner way if possible.
-        // At the very least some of this is overkill.
-        $variable_names = $this->getTwigVariableNames($node, array_replace_recursive($all_variables, $variables), $set_var, $set_source);
-        $variable_names_flat = [];
-        $prev_name = null;
-        foreach ($variable_names as $name) {
-          $name_check = $prev_name ? $prev_name['path'] . '.' . $name['path']:'';
-          if (($name['path'] == $set_var) && !($set_source == $name_check)) {
-            $variable_names_flat[] = $set_source;
-          }
-          else {
-            $variable_names_flat[] = $name['path'];
-          }
-          $prev_name = $name;
-        }
-        $variable_key = implode('.', $variable_names_flat);
-      }
-      elseif ($node instanceof Node) {
-        // See above comment about the recursion.
-        $add_variables = $this->getTwigVariableNames($node, array_replace_recursive($all_variables, $variables), $set_var, $set_source);
-        $variables = array_replace_recursive($variables, $add_variables);
-      }
-      if (!empty($variable_key)) {
+      if (!empty($variable_key && str_starts_with($variable_key, 'data.'))) {
         $variables[$variable_key]['path'] = $variable_key;
-        if(isset($all_variables[$variable_key]['line'])) {
-          // Since $variables is lost with each return and $all_variables accumulates
-          // line numbers across recursions we need to merge the current line
-          // number with those passed thru recursions, and since duplicate
-          // values can occur because it's an array, we need unique values.
-          $variables[$variable_key]['line']  = array_unique(array_merge($all_variables[$variable_key]['line'], $lineno));
-        }
-        if(!isset($variables[$variable_key]['line'])) {
-          $variables[$variable_key]['line'] = $lineno;
-        }
-        if (!empty($parent_path)) {
-          $variables[$variable_key]['parent_path'] = $parent_path;
+        $variables[$variable_key]['line'] = isset($variables[$variable_key]['line'])
+          ? array_merge($variables[$variable_key]['line'], $lineno)
+          : $lineno;
+      }
+      if ($node instanceof Node) {
+        $add_variables = $this->getTwigVariableNames($node);
+        if (!empty($variables) || !empty($add_variables)) {
+          $variables = static::arrayMergeRecursive($variables, $add_variables);
```

Comment:
There is excessive recursion here. Basically the arrayMergeRecursive will traverse over and over the same keys (said differently, If the array has N entries, you will traverse N times the last loop, N-1 time the almost last, N-2, etc), when you could really just traverse N times or not at all using https://github.com/esmero/format_strawberryfield/pull/351/files#diff-b5161fcfb4c915fb64182570502fc39348dd9af67772fa46c6e2a7f24f814e3eR553 and passing $variables as a/by reference argument between recursions.